metrics/config.go

```diff
@@ -0,0 +1,200 @@
+)
+
+var (
+	exporter    view.Exporter
+	mConfig     metricsConfig
+	mux         sync.Mutex
+	promSrvChan chan *http.Server = make(chan *http.Server, 1)
+)
+
+// newMetricsExporter is a blocking operation to get a metrics exporter based on the config.
+func newMetricsExporter(config metricsConfig, logger *zap.SugaredLogger) error {
+	var err error
+	mux.Lock()
+	defer mux.Unlock()
+	select {
+	case svr := <-promSrvChan:
+		svr.Close()
+	default:
+	}
+
+	if exporter != nil {
+		view.UnregisterExporter(exporter)
+	}
+	if config.backendDestination == Stackdriver {
+		exporter, err = newStackdriverExporter(config, logger)
+	} else {
+		exporter, err = newPrometheusExporter(config, logger)
+	}
+	if err != nil {
+		return err
+	}
+	view.RegisterExporter(exporter)
+	view.SetReportingPeriod(10 * time.Second)
+	logger.Infof("Successfully updated the metrics exporter; old config: %v; new config %v", mConfig, config)
+	mConfig = config
+	return nil
+}
+
+func newStackdriverExporter(config metricsConfig, logger *zap.SugaredLogger) (view.Exporter, error) {
+	e, err := stackdriver.NewExporter(stackdriver.Options{
+		ProjectID:    config.stackdriverProjectID,
+		MetricPrefix: config.domain + "/" + config.component,
+		Resource: &monitoredrespb.MonitoredResource{
+			Type: "global",
+		},
+		DefaultMonitoringLabels: &stackdriver.Labels{},
+	})
+	if err != nil {
+		logger.Error("Failed to create the Stackdriver exporter.", zap.Error(err))
+		return nil, err
+	}
+	logger.Info("Created Opencensus Stackdriver exporter with config:", config)
+	return e, nil
+}
+
+func newPrometheusExporter(config metricsConfig, logger *zap.SugaredLogger) (view.Exporter, error) {
+	e, err := prometheus.NewExporter(prometheus.Options{Namespace: config.component})
+	if err != nil {
+		logger.Error("Failed to create the Prometheus exporter.", zap.Error(err))
+		return nil, err
+	}
+	logger.Info("Created Opencensus Prometheus exporter with config:", config)
+	logger.Info("Start the endpoint for Prometheus exporter.")
+	// Start the endpoint for Prometheus scraping
+	go func() {
+		sm := http.NewServeMux()
+		sm.Handle("/metrics", e)
+		promSrv := &http.Server{
+			Addr:    ":9090",
+			Handler: sm,
+		}
+		promSrvChan <- promSrv
+		promSrv.ListenAndServe()
+	}()
+	return e, nil
+}
+
+func getMetricsConfig(m map[string]string, domain string, component string, logger *zap.SugaredLogger) (metricsConfig, error) {
+	var mc metricsConfig
+	backend, ok := m[backendDestinationKey]
+	if !ok {
+		return mc, errors.New("metrics.backend-destination key is missing")
+	}
+	var mb MetricsBackend
+	if strings.EqualFold(backend, "stackdriver") {
+		mb = Stackdriver
+	} else {
+		mb = Prometheus
+		if !strings.EqualFold(backend, "prometheus") {
+			logger.Infof("Unsupported metrics backend value \"%s\". Use the default metrics backend prometheus.", backend)
+		}
+	}
+	mc.backendDestination = mb
+
+	sdProj, ok := m[stackdriverProjectIDKey]
+	if strings.EqualFold(backend, "stackdriver") && !ok {
```

Comment:
why not mb.backendDestination == Stackdriver? Would be cheaper to do and easier to understand. Also, following might look better (two ifs combined):
```
if mc.backendDestination == Stackdriver {
    sdProj, ok := m[stackdriverProjectIDKey]
    if !ok || sdProj == "" {
        return mc, some error
    }
}
```